functions/v2/typed/googlechatapp/main.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# [START functions_typed_googlechatbot]
+import functions_framework
+
+
+class ChatRequest:
+    message: dict
+
+    # Required to deserialize the request
+    @staticmethod
+    def from_dict(req: dict) -> "ChatRequest":
```

Comment:
The forward declaration could work, however does not actually do much of its job for enforcing the type when the program is actually run.

Instead, use `ChatRequest` type hint directly, and don't forget to import `from __future__ import annotations` - this is default behavior starting in 3.10 but we'll need to account for <3.10 versions being used.

Reply:
Adding the `from __future__ import annotations` seems to actually be changing the way python's reflection APIs work which is pretty unexpected.

We can test this with a snippet that looks like 

```
from inspect import signature

def myfunc(req: GreetingRequest):
    pass

sig = signature(myfunc)
annotation_type = list(sig.parameters.values())[0].annotation
print(type(annotation_type))
```

When `from __future__ import annotations` is included the output is `str`, the and the value is the string `GreetingRequest`.

When the future annotations feature is omitted, the output is `<class 'type'>` and the value is `<class 'main.GreetingRequest'>`. 

The typed function feature relies on reflection on the declared function type to determine the input/output and it depends on the reflection APIs returning the actual class of the request / response. I don't think I can bring in `from __future__ import annotations` but this does worry me that this feature could break when python adopts the new reflection behavior.

Reply:
>  relies on reflection on the declared function type to determine the input/output and it depends on the reflection APIs returning the actual class of the request / response

This might be a side effect of using `from __future__ import annotations` which I believe does post-processing rather than pre-processing that could prematurely think that the class object does not exist, although it's being defined as the type checker is running through. Hence, it might look like the type has changed or looks weird / dynamic but the end result should be accurate.

If we use forward declaration as is, we just visually know that the type returned should be a class object but you could pass in a string or another object and it would only fail at runtime if the code has run into an exception.

How does adding it actually change the way that the API works from this code? It just seems like we're missing precise type definitions (you also need to use `from typing import Dict` instead of just `dict` until I believe 3.8 or 3.9) rather than API behavior confusion as you describe it